builtin/credential/approle/path_role.go

```diff
@@ -857,55 +857,79 @@ func (b *backend) pathRoleRead(req *logical.Request, data *framework.FieldData)
 
 	lock := b.roleLock(roleName)
 	lock.RLock()
-	defer lock.RUnlock()
+	writeLockHeld := false
 
-	if role, err := b.roleEntry(req.Storage, strings.ToLower(roleName)); err != nil {
+	role, err := b.roleEntry(req.Storage, strings.ToLower(roleName))
+	if err != nil {
+		lock.RUnlock()
 		return nil, err
-	} else if role == nil {
+	}
+
+	if role == nil {
+		lock.RUnlock()
 		return nil, nil
-	} else {
-		// Convert the 'time.Duration' values to second.
-		role.SecretIDTTL /= time.Second
-		role.TokenTTL /= time.Second
-		role.TokenMaxTTL /= time.Second
-		role.Period /= time.Second
+	}
 
-		// Create a map of data to be returned and remove sensitive information from it
-		data := structs.New(role).Map()
-		delete(data, "role_id")
-		delete(data, "hmac_key")
+	// Convert the 'time.Duration' values to second.
```

Comment:
If we switch to a write lock, this will now persist the wrong values. You can just set the resp data explicitly.

Reply:
Note that we are not persisting the role. We are only creating an index for the role ID. But still we hold the write lock on the role since the index is tightly coupled to a role entry.